garb/garb_main.cpp

```diff
@@ -224,14 +224,162 @@ main (int argc, char* argv[])
+        assert(mutex != nullptr);
+
+        gu_mutex_destroy (mutex);
+        delete mutex;
+        *value = nullptr;
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_LOCK: {
+        gu_mutex_t *mutex = reinterpret_cast<gu_mutex_t *>(*value);
+        assert(mutex != nullptr);
+
+        gu_mutex_lock(mutex);
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_UNLOCK: {
+        gu_mutex_t *mutex = reinterpret_cast<gu_mutex_t *>(*value);
+        assert(mutex != nullptr);
+
+        gu_mutex_unlock(mutex);
+
+        break;
+      }
+
+      default:
+        assert(0);
+        break;
+    }
+  } else if (type == WSREP_PFS_INSTR_TYPE_CONDVAR) {
+    assert (value != nullptr);
+    switch (ops) {
+      case WSREP_PFS_INSTR_OPS_INIT: {
+        gu_cond_t *cond = new gu_cond_t();
+
+        gu_cond_init(cond, nullptr);
+        *value = cond;
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_DESTROY: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
+        assert(cond != nullptr);
+
+        gu_cond_destroy(cond);
+        delete cond;
+        *value = nullptr;
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_WAIT: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
+        gu_mutex_t *mutex = reinterpret_cast<gu_mutex_t *>(*alliedvalue);
+        assert(cond != nullptr);
+
+        gu_cond_wait (cond, mutex);
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_TIMEDWAIT: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
+        gu_mutex_t *mutex = reinterpret_cast<gu_mutex_t *>(*alliedvalue);
+        const timespec *wtime = reinterpret_cast<const timespec *>(ts);
+        assert(cond != nullptr && mutex != nullptr);
+
+        gu_cond_timedwait(cond, mutex, wtime);
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_SIGNAL: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
+        assert(cond != nullptr);
+
+        gu_cond_signal(cond);
+
+        break;
+      }
+
+      case WSREP_PFS_INSTR_OPS_BROADCAST: {
+        gu_cond_t *cond = reinterpret_cast<gu_cond_t *>(*value);
+        assert(cond != nullptr);
+
+        gu_cond_broadcast(cond);
+
+        break;
+      }
+
+      default:
+        assert(0);
+        break;
+    }
+  } else if (type == WSREP_PFS_INSTR_TYPE_THREAD) {
+    switch (ops) {
+      case WSREP_PFS_INSTR_OPS_INIT:
+      case WSREP_PFS_INSTR_OPS_DESTROY:
+        break;
+
+      default:
+        assert(0);
+        break;
+    }
+  } else if (type == WSREP_PFS_INSTR_TYPE_FILE) {
+    switch (ops) {
+      case WSREP_PFS_INSTR_OPS_CREATE:
+      case WSREP_PFS_INSTR_OPS_OPEN:
+      case WSREP_PFS_INSTR_OPS_CLOSE:
+      case WSREP_PFS_INSTR_OPS_DELETE:
+        break;
+
+      default:
+        assert(0);
+        break;
+    }
+  }
+}
 
 int
 main (int argc, char* argv[])
 {
-    gu_conf_set_pfs_instr_callback(pfs_noop);
+    gu_conf_set_pfs_instr_callback(dummy_pfs_cb);
```

Comment:
instead of providing dummy_pfs_cb? can we avoid registering this callback? what would happen? :thinking:

Reply:
Not a good idea. In case of compilation with `-DHAVE_PSI_INTERFACE=1` the surrounding code will call nullptr callback. And now by default we compile with `-DHAVE_PSI_INTERFACE=1` as it is needed by libgalera_smm.so (see changes in cmake files part)
In case of compilation without -DHAVE_PSI_INTERFACE=1, nothing SHOULD happen as the compiler should fall back to plain mutex/condition objects and this callback will be never called.
But prior to change we registered an empty callback always, so let's play safe.

Reply:
agree